forte/data/data_store.py

```diff
@@ -490,6 +492,72 @@ def get_entry(self, tid: int) -> Tuple[List, int, int]:
         # Otherwise, use ``index_id`` to find the index of the entry.
         raise NotImplementedError
 
+    def co_iterator(self, type_names: List[str]) -> Iterator[List]:
+        r"""Given two or more type names, iterate their entry lists from beginning to end together.
+
+        For every single type, their entry lists are sorted by the begin and
+        end fields. The co_iterator function will iterate those sorted lists
+        together, and yield each entry in sorted order. This tasks is quite
+        similar to merging several sorted list to one sorted list. If two
+        entries have both the same begin and end field, then their order is
+        decided by the order of user input type_name (the type that first
+        appears in the target type list will return first).
+
+        Args:
+            type_names (List[str]): a list of string type names
+
+        Returns:
+
+            An iterator of entry elements.
+        """
+
+        n = len(type_names)
+        # suppose the length of type_names is N and the length of entry list of
+        # one type is M
+        # then the time complexity of using min-heap to iterate
+        # is O(M*log(N))
+
+        entries = []
+        for tn in type_names:
+            if tn in self.__elements:
+                entries.append(self.__elements[tn][0])
+            else:
+                raise ValueError(
+                    f"Input parameter types name {tn} is not"
+                    "available. Please input avaible ones in this DataStore"
+                    f"object: {list(self.__elements.keys())}"
+                )
+        pointers = [0] * n  # record the current entry index for elements
+        # compare tuple (begin, end, order of type name in input argument type_names)
+        h = []
+        begin_idx, end_idx, type_name_idx = 0, 1, 3
+        for p_idx in range(n):
+            h.append(
+                (
+                    (entries[p_idx][begin_idx], entries[p_idx][end_idx], p_idx),
+                    entries[p_idx][type_name_idx],
+                )
+            )  # p_idx
+
+        heapify(h)
```

Comment:
Can you add some comments explaining what the heap is doing? And btw, could we build the heap on the fly instead of doing it later.